### PR TITLE
apr 20 notes

### PR DESCRIPTION
**Pull request by YafuR**

### CODE REVIEW

File: apr20/yr942.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+As disscussed in the paper, drop out is a popular technique to avoid overfitting and acceleration, especially in deep neural network. For local environment, it is easy to accomplish. However, during DDP it could not save time since fixed parameter-to-bucket mapping. Consequently, regardless of how the forward pass skips layers, there is always the same amount of data to be communicated across the wire during the backward pass.
+
+- Compressoin: By compressing the gradient, the volume of data during communication could be much less.
+
+
+
+# TVM
+## Motivation
+
+Currently there are various of hardware accelerators and we need to rebuild the entire software stack on top of it for each of them. So it't better if we could design a bridge between the user's customized framework and the hardware.
```
Comment (kmish0923) [inline apr20/yr942.md]: typo: various hardware*

Comment (kmish0923) [reply apr20/yr942.md]: it'd*

File: apr20/yr942.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+- The operator-level optimization module must generate efficient code for each fused operator in this graph: TVM produces efficient code for each operator by generating many valid implementations on each hardware back-end and choosing an optimized implementation.
+- We use use an ML-based cost model to find optimized operators.
+- The system packs the generated code into a deployable module.
+
+
+## Open Questions and Future Work
+
+The co-design between the software and hardware, so more future work may be based on it since we will need different accelerators for models.
+
+
```
Comment (kmish0923) [inline apr20/yr942.md:90]: Tradeoffs?

File: apr20/yr942.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+## Trade-Offs
+
+- Clipper does not optimize the model within the frameworks.
+- Clipper does not manage the training or re-training model within the frameworks.
+
+## Open Questions and Future Work
+
+The clipper could not improve the performace of the model if it is out-of-date. Could we do some interaction between clipper and models to improve the accuracy?
```
Comment (kmish0923) [inline apr20/yr942.md:135]: What do you mean out of date?

File: apr20/yr942.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+2. Gradient Reduction
+
+- Naive solution: PyTorch autograd engine accepts custom backward hooks. DDP can register autograd hooks to trigger computation after every backward pass. When fired, each hook scans through all local model parameters, and retrieves the gradient tensor from each parameter. Then, it uses the AllReduce collective communication call to calculate the average gradients on each parameter across all processes, and writes the result back to the gradient tensor.
+
+- Gradient Bucketing: Experients shows that the total communication time decreases when using larger input tensors. Instead of launching a dedicated AllReduce immediately when each gradient tensor becomes available, DDP can achieve higher throughput and lower latency if it waits for a short period of time and buckets multiple gradients into one AllReduce operation.
+
+- Overlap Computation with Communication
+
+- Gradient Accumulation: Here AllReduce is not called in each iteration. Instead, the application could do n iterations before gradient synchronizing.
```
Comment (neeraj71) [inline apr20/yr942.md:43]: I am wondering why pytorch require gradient accumulation if they are already doing bucketing?

File: apr20/yr942.md (hunk 3 above)
Comment (YafuR) [reply apr20/yr942.md:135]: The clipper could not retrain the model if the enviromnent changes, and in that case I mean those models are out of date.

Comment (YafuR) [reply apr20/yr942.md:135]: Also, I don't think clipper could deal with online machining learnning in which case the learning environment changes fast.

File: apr20/yr942.md (hunk 4 above)
Comment (YafuR) [reply apr20/yr942.md:43]: In my understanding, bucketing means we are packing all gradients in one iteration. For example, for a neural network we could have thousands of weights and bias, and for each of them we have one gradient. We could set buckets n < # of parameters to make the tensor large, which could improve the communication efficiency.  Gradient accumulation means we do several iterations before we synchronizing gradients globally, and in this case I guess we are reducing those buckets.

File: apr20/yr942.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+
+# Clipper
+## Motivation
+
+There are a lot of systems designed for training models. However the systems designed for serving the application after the model has been trained, is much less. There are several challenges when we design this system:
+- It need to support low-latency, high througput serving workloads.
+- It need to support different ML ecosystems and frameworks.
+
+There are existing serving system, trying to make all possible predictions and store them in to query offline. During serving, the application just look up the revelant query in database. However the application is bounded into the the pre-disigned database.
```
Comment (li928904115) [inline apr20/yr942.md:98]: What do you mean "bounded by pre-designed database" here?

File: apr20/yr942.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+The Model Selection Layer uses feedback to dynamically select one or more of the deployed models and combine their outputs to provide more accurate and robust predictions.
+- Exploit multiple models to estimate confidence
+- Use multi-armed bandit algorithms to learn optimal model-selection online
+- Online personalization across ML frameworks
+
+
+## Trade-Offs
+
+- Clipper does not optimize the model within the frameworks.
```
Comment (li928904115) [inline apr20/yr942.md:130]: Can you further explain what does this mean?

Comment (vidit23) [reply apr20/yr942.md:130]: I can answer that, Clipper is just an inference delivery system. Thus, the models are treated like black boxes. However, each model's performance is highly dependent on how well it runs on the underlying hardware (TVM does this). If the model is not optimized for the hardware, it'll run slowly despite how much Clipper tries.

File: apr20/yr942.md (hunk 5 above)
Comment (YafuR) [reply apr20/yr942.md:98]: For some traditional serving system, it stores some pre-calculated features and predictions in database. When application requests, it searches through the query to match the features. However, the database could only store discrete values, which means the application only get approximate predictions instead of exact ones. That's what I mean "bounded".
